pkg/api/validation/events.go

```diff
@@ -25,13 +29,32 @@ import (
 // ValidateEvent makes sure that the event makes sense.
 func ValidateEvent(event *api.Event) errs.ValidationErrorList {
 	allErrs := errs.ValidationErrorList{}
-	// TODO: There is no namespace required for node.
-	if event.InvolvedObject.Kind != "Node" &&
-		event.Namespace != event.InvolvedObject.Namespace {
-		allErrs = append(allErrs, errs.NewFieldInvalid("involvedObject.namespace", event.InvolvedObject.Namespace, "namespace does not match involvedObject"))
+
+	nonNamespacedKindFlag, err := isNonNamespacedKind(event.InvolvedObject.Kind, event.InvolvedObject.APIVersion)
+	if err != nil {
+		allErrs = append(allErrs, errs.NewFieldInvalid("involvedObject.Kind", event.InvolvedObject.Kind, fmt.Sprintf("couldn't check whether namespace is allowed: %s", err)))
+	} else {
+		if !nonNamespacedKindFlag &&
+			event.Namespace != event.InvolvedObject.Namespace {
+			allErrs = append(allErrs, errs.NewFieldInvalid("involvedObject.namespace", event.InvolvedObject.Namespace, "namespace does not match involvedObject"))
+		}
 	}
 	if !validation.IsDNS1123Subdomain(event.Namespace) {
 		allErrs = append(allErrs, errs.NewFieldInvalid("namespace", event.Namespace, ""))
 	}
 	return allErrs
 }
+
+// Check whether the kind in apiVersion is scoped at the root of the api hierarchy
+func isNonNamespacedKind(kind, groupVersion string) (bool, error) {
+	group := apiutil.GetGroup(groupVersion)
+	restMapping, err := latest.GroupOrDie(group).RESTMapper.RESTMapping(kind, groupVersion)
```

Comment:
GroupOrDie is OK if the argument passed to it is constant, i.e. "", "v1",
"extensions".

On Wed, Oct 28, 2015 at 11:52 PM, Flying Tiger <notifications@github.com>
wrote:

> In pkg/api/validation/events.go
> <https://github.com/kubernetes/kubernetes/pull/16146#discussion_r43353767>
> :
>
> >  	}
> >  	if !validation.IsDNS1123Subdomain(event.Namespace) {
> >  		allErrs = append(allErrs, errs.NewFieldInvalid("namespace", event.Namespace, ""))
> >  	}
> >  	return allErrs
> >  }
> > +
> > +// Check whether the kind in apiVersion is scoped at the root of the api hierarchy
> > +func isNonNamespacedKind(kind, groupVersion string) (bool, error) {
> > +	group := apiutil.GetGroup(groupVersion)
> > +	restMapping, err := latest.GroupOrDie(group).RESTMapper.RESTMapping(kind, groupVersion)
>
> Fixed this, please review. But I found GroupOrDie() is much more used than
> Group(). So I added a todo note to strengthen abandonment of this function.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16146/files#r43353767>.
>